solver/src/naive_solver/multi_order_solver.rs

```diff
@@ -0,0 +1,369 @@
+        (context_a, context_b)
+    };
+    let uniswap_out = compute_uniswap_out(&shortage, &excess);
+    let uniswap_in = compute_uniswap_in(uniswap_out, &shortage, &excess);
+    let interaction = if !uniswap_out.is_zero() {
+        Some(AmmSwapExactTokensForTokens {
+            amount_in: uniswap_in,
+            amount_out_min: uniswap_out,
+            token_in: excess.address,
+            token_out: shortage.address,
+        })
+    } else {
+        // TODO(fleupold) set correct clearing prices when supply/demand already match current uniswap price
+        None
+    };
+    // TODO(fleupold) check that all orders comply with the computed price. Otherwise, remove the least favorable excess order and try again.
+    SinglePairSettlement {
+        clearing_prices: maplit::hashmap! {
+            shortage.address => uniswap_in,
+            excess.address => uniswap_out,
+        },
+        trades: orders.into_iter().map(Trade::fully_matched).collect(),
+        interaction,
+    }
+}
+
+fn split_into_contexts(
+    orders: impl Iterator<Item = OrderCreation>,
+    reserves: &HashMap<Address, U256>,
+) -> (TokenContext, TokenContext) {
+    let mut contexts = HashMap::new();
+    for order in orders {
+        let buy_context = contexts
+            .entry(order.buy_token)
+            .or_insert_with(|| TokenContext {
+                address: order.buy_token,
+                reserve: *reserves
+                    .get(&order.buy_token)
+                    .unwrap_or_else(|| panic!("No reserve for token {}", &order.buy_token)),
+                ..Default::default()
+            });
+        if matches!(order.kind, OrderKind::Buy) {
+            buy_context.buy_volume += order.buy_amount
+        }
+
+        let sell_context = contexts
+            .entry(order.sell_token)
+            .or_insert_with(|| TokenContext {
+                address: order.sell_token,
+                reserve: *reserves
+                    .get(&order.sell_token)
+                    .unwrap_or_else(|| panic!("No reserve for token {}", &order.sell_token)),
+                ..Default::default()
+            });
+        if matches!(order.kind, OrderKind::Sell) {
+            sell_context.sell_volume += order.sell_amount
+        }
+    }
+    assert!(contexts.len() == 2, "Orders contain more than two tokens");
+    let mut contexts = contexts.drain().map(|(_, v)| v);
+    (contexts.next().unwrap(), contexts.next().unwrap())
+}
+
+/**
+ * Given information about the shortage token (the one we need to take from Uniswap) and the excess token (the one we give to Uniswap), this function
+ * computes the exact out_amount required from Uniswap to perfectly match demand and supply at the effective Uniswap price (the one used for that in/out swap).
+ *
+ * The derivation of this formula is described in https://docs.google.com/document/d/1jS22wxbCqo88fGsqEMZgRQgiAcHlPqxoMw3CJTHst6c/edit
+ * It assumes GP fee (φ) to be 1 and Uniswap fee (Φ) to be 0.997
```

Comment:
I'm not sure this is right... or does this assume that a fee discount will be computed to adjust for the Uniswap trade?

Reply:
This is basically assuming there is no GP related fee (only the 0.997 uniswap fee). As for the real solver, I think the fee can be deducted from the sell/buy amounts before the orders are passed into the solver engine. The solvers themselves can run on the assumption that there is no protocol fee.

Reply:
Ah OK, I remember the decision now, we would pass the instance without the fees (deducted from the balances and not amounts I think right?) and then the driver would adjust the `feeDiscount` to account for any Uniswap trades. In this way, the solver and the naive solver can share the same code for computing fee discounts.